tests/thermodynamics/test_hessianJob.py

```diff
@@ -49,5 +49,33 @@ def test_forces(self):
         self.assertAlmostEqual(self.job.output.pressures[0, 0, 0], -0.03)
 
 
+class TestSelectiveDynamics(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.file_location = os.path.dirname(os.path.abspath(__file__))
+        cls.project = Project(
+            os.path.join(cls.file_location, "selective_dynamics")
+        )
+        cls.project.remove_jobs_silently(recursive=True)
+
+    def test_selective_dynamics(self):
+        structure = Atoms(
+            positions=[[0, 0, 0], [1, 1, 1]], elements=["Fe", "Fe"], cell=2 * np.eye(3)
+        )
+        structure.add_tag(selective_dynamics=None)
+        structure.selective_dynamics[0] = [True, True, True]
+        structure.selective_dynamics[1] = [False, False, False]
+        job = self.project.create_job("HessianJob", "hess")
+        job.structure = structure
+        job.to_hdf()
+        job_reload = self.project.create_job("HessianJob", "hess")
+        job_reload.from_hdf()
+        structure_reload = job_reload.structure
+        structure_reload.add_tag(selective_dynamics=None)
+        structure_reload.selective_dynamics[0] = [True, True, True]
+        structure_reload.selective_dynamics[1] = [False, False, False]
+        self.assertTrue(True)
```

Comment:
This doesn't do anything right? Maybe we can check the change in positions to see if the atoms have moved or not.

Reply:
The issue is that without the fix reloading the selective dynamics tag from the HDF5 file gives the following error: 
```
======================================================================
ERROR: test_selective_dynamics (thermodynamics.test_hessianJob.TestSelectiveDynamics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/tests/thermodynamics/test_hessianJob.py", line 76, in test_selective_dynamics
    structure_reload.selective_dynamics[1] = [False, False, False]
  File "/home/vsts/work/1/s/pyiron/atomistics/structure/sparse_list.py", line 203, in __setitem__
    raise IndexError
IndexError
```
Therefore this test basically just confirms that the reloading works and the selective dynamics can be overwritten afterwards.

Reply:
Maybe it's clearer to do the `from_hdf` in an `try/except IndexError` and call `self.fail` in the except clause to indicate test failure.

Reply:
The bug is now fixed - I do not want it to fail - I just say it was failing before, as the length of the tag array was not restored correctly.

Reply:
> The issue is that without the fix reloading the selective dynamics tag from the HDF5 file gives the following error:
> 
> ```
> ======================================================================
> ERROR: test_selective_dynamics (thermodynamics.test_hessianJob.TestSelectiveDynamics)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/vsts/work/1/s/tests/thermodynamics/test_hessianJob.py", line 76, in test_selective_dynamics
>     structure_reload.selective_dynamics[1] = [False, False, False]
>   File "/home/vsts/work/1/s/pyiron/atomistics/structure/sparse_list.py", line 203, in __setitem__
>     raise IndexError
> IndexError
> ```
> 
> Therefore this test basically just confirms that the reloading works and the selective dynamics can be overwritten afterwards.

Then you should do something like `job["output/structure"].to_object()` and print it's selective dynamics values to check.

Reply:
I understand that the test don't fail after the other PR, but my reasoning is that failing explicitly in case of `IndexError` makes it easier to identify a regression in the  future.  I'm also happy with the more explicit asserts from @sudarsan-surendralal, though.